src/cpp/session/modules/SessionDataImportV2.R

```diff
@@ -13,21 +13,180 @@
+        sep = " = ")
+    }
+  }
+
+  spacing <- ",\n  "
+  paste(parameters, collapse = spacing)
+})
+
+.rs.addFunction("assemble_data_import", function(dataImportOptions)
+{
+  importInfo <- list()
+
+  dataName <- importInfo$dataName <- .rs.assemble_data_import_name(dataImportOptions)
+  if (is.null(dataName) || identical(dataName, ""))
+  {
+    dataName <- "dataset"
+  }
+
+  if (!grepl("^[a-zA-Z]+[a-zA-Z0-9_]*$", dataName) || any(dataName == ls('package:base')))
+  {
+    dataName <- paste("`", dataName, "`", sep = "")
+  }
+
+  functionName <- list()
+  functionReference <- readr::read_delim
+  if (is.null(dataImportOptions$delimiter) || identical(dataImportOptions$delimiter, ","))
+  {
+    functionName <- "read_csv"
+    functionReference <- readr::read_csv
+  }
+  else
+  {
+    functionName <- "read_delim"
+  }
+
+  # load parameters
+  options <- list()
+  options[["file"]] <- dataImportOptions$importLocation
+  options[["delim"]] <- dataImportOptions$delimiter
+  options[["quote"]] <- dataImportOptions$quotes
+  options[["escape_backslash"]] <- dataImportOptions$escapeBackslash
+  options[["escape_double"]] <- dataImportOptions$escapeDouble
+  options[["col_names"]] <- dataImportOptions$columnNames
+  options[["trim_ws"]] <- dataImportOptions$trimSpaces
+  options[["locale"]] <- dataImportOptions$locale
+  options[["na"]] <- dataImportOptions$na
+  options[["comment"]] <- dataImportOptions$comments
+  options[["skip"]] <- dataImportOptions$skip
+  options[["n_max"]] <- dataImportOptions$maxRows
+
+  # set special parameter types
+  optionTypes <- list()
+  optionTypes[["file"]] <- "character"
+  optionTypes[["delim"]] <- "character"
+  optionTypes[["quote"]] <- "character"
+  optionTypes[["locale"]] <- "locale"
+  optionTypes[["na"]] <- "character"
+  optionTypes[["comment"]] <- "character"
+
+  functionParameters <- .rs.assemble_data_import_parameters(options, optionTypes, functionReference)
+
+  importInfo$previewCode <- paste(
+    "readr::",
+    functionName,
+    "(",
+    functionParameters,
+    ")",
+    sep = "")
+
+  importInfo$importCode <- paste(dataName, " <- ", importInfo$previewCode, "\n", "View(", dataName, ")", sep = "")
+
+  importInfo
+})
+
+.rs.addJsonRpcHandler("assemble_data_import", function(dataImportOptions)
 {
   tryCatch({
-    data <- readr::read_csv(path, n_max = maxRows)
+    return (.rs.assemble_data_import(dataImportOptions))
+  }, error = function(e) {
+    return(list(error = e))
+  })
+})
+
+.rs.addJsonRpcHandler("preview_data_import", function(dataImportOptions, maxCols = 100, maxFactors = 64)
+{
+  tryCatch({
+
+    importInfo <- .rs.assemble_data_import(dataImportOptions)
+    data <- eval(parse(text=importInfo$previewCode))
     columns <- .rs.describeCols(data, maxCols, maxFactors)
 
     cnames <- names(data)
     size <- nrow(data)
 
-    for(cname in cnames[-1]) {
-      data[[cname]] <- .rs.formatDataColumn(data[[cname]], 1, size)
+    for(i in 1:ncol(data)) {
```

Comment:
You might want to use `seq_len(ncol(data))` instead; in the edge case of a zero-column dataset you'll get the vector `[1, 0]` which is probably not what you want.

You might also just want `for (i in seq_along(data)) { ... }`; `seq_along()` will generate an index vector up to the length of the data.